wallet/src/Pos/Wallet/Web/Methods/History.hs

```diff
@@ -76,45 +75,45 @@ getFullWalletHistory cWalId = do
     addHistoryTxs cWalId localHistory
     logDebug "getFullWalletHistory: invoked addHistoryTxs"
 
-    !cHistory <- forM fullHistory (constructCTx cWalId walAddrsDetector diff)
-    logDebug "getFullWalletHistory: formed cTxs"
-    pure (cHistory, fromIntegral $ Map.size cHistory)
+    pure (fullHistory, fromIntegral $ Map.size fullHistory)
 
-getHistory
-    :: MonadWalletWebMode m
+getFilteredHistory
+    :: forall m . MonadWalletWebMode m
     => CId Wal
     -> [AccountId]
     -> Maybe (CId Addr)
-    -> m (Map TxId (CTx, POSIXTime), Word)
-getHistory cWalId accIds mAddrId = do
+    -> m (Map TxId TxHistoryEntry, Word)
+getFilteredHistory cWalId accIds mCAddrId = do
     -- FIXME: searching when only AddrId is provided is not supported yet.
-    accAddrs  <- S.fromList . map (cwamId . adiCWAddressMeta) <$> concatMapM (getAccountAddrsOrThrow Ever) accIds
+    accCAddrs <- map (cwamId . adiCWAddressMeta) <$> concatMapM (getAccountAddrsOrThrow Ever) accIds
     allAccIds <- getWalletAccountIds cWalId
 
-    let filterFn :: Map TxId (CTx, POSIXTime) -> Map TxId (CTx, POSIXTime)
-        !filterFn = case mAddrId of
+    let !filterFn = case mCAddrId of
           Nothing
             | S.fromList accIds == S.fromList allAccIds
               -- can avoid doing any expensive filtering in this case
-                        -> identity
-            | otherwise -> filterByAddrs accAddrs
+                        -> pure
+            | otherwise -> \history -> do
+                  accAddrs <- S.fromList <$> convertCIdToAddrs accCAddrs
+                  pure $ filterByAddrs accAddrs history
 
-          Just addr
-            | addr `S.member` accAddrs -> filterByAddrs (S.singleton addr)
+          Just caddr
+            | caddr `elem` accCAddrs -> \history -> do
+                 addr <- convertCIdToAddr caddr
+                 pure $ filterByAddrs (S.singleton addr) history
             | otherwise                -> throw errorBadAddress
-
-    res <- first filterFn <$> getFullWalletHistory cWalId
-    logDebug "getHistory: filtered transactions"
-    return res
+    res <- (\(x, y) -> (,y) <$> filterFn x) =<< getFullWalletHistory cWalId
+    res <$ logDebug "getFilteredHistory: filtered transactions"
```

Comment:
The previous way of writing this is much more straightforward. Is there a good reason for rewriting it?